In [2]:
import pandas as pd
import math

In [3]:
df = pd.read_json("./experiment_queries_execution_no_rerank2.json", orient="records")
df
    

method_id  user_id  class_id  \
0                                    single_elastic_bm25        1  20825628   
1                                    single_elastic_bm25        1  20825628   
2                                    single_elastic_bm25        1  20825628   
3                                    single_elastic_bm25        1  20825628   
4                                    single_elastic_bm25        1  20825628   
...                                                  ...      ...       ...   
99523  tuple_fusion_qdrant_sparse_0.9_x_elastic_bm25_0.1      114     23442   
99524  tuple_fusion_qdrant_sparse_0.9_x_elastic_bm25_0.1      114     23442   
99525  tuple_fusion_qdrant_sparse_0.9_x_elastic_bm25_0.1      114     23442   
99526  tuple_fusion_qdrant_sparse_0.9_x_elastic_bm25_0.1      114     23442   
99527  tuple_fusion_qdrant_sparse_0.9_x_elastic_bm25_0.1      114     23442   

                      query_id               class_category    rr      time  
0                 user_query_0           instances_100_1000  0.00  0.115999  
1                 user_query_1           instances_100_1000  0.00  0.069224  
2                 user_query_2           instances_100_1000  0.00  0.050538  
3                 user_query_3           instances_100_1000  0.00  0.081411  
4                 user_query_4           instances_100_1000  0.00  0.255341  
...                        ...                          ...   ...       ...  
99523  mech_label_user_query_1  instances_100000_1000000000  1.00  0.088526  
99524  mech_label_user_query_2  instances_100000_1000000000  1.00  0.079080  
99525  mech_label_user_query_3  instances_100000_1000000000  0.50  0.084768  
99526  mech_label_user_query_4  instances_100000_1000000000  0.25  0.089515  
99527  mech_label_user_query_5  instances_100000_1000000000  1.00  0.080565  

[99528 rows x 7 columns]

In [4]:
# Number of methods
df["method_id"].nunique()

66

# Overall

In [5]:
df_thin = df[["method_id", "rr", "time"]]
df_thin.head()

method_id   rr      time
0  single_elastic_bm25  0.0  0.115999
1  single_elastic_bm25  0.0  0.069224
2  single_elastic_bm25  0.0  0.050538
3  single_elastic_bm25  0.0  0.081411
4  single_elastic_bm25  0.0  0.255341

In [6]:
df_overall =  df_thin.groupby(["method_id"], as_index=False).mean().sort_values(by="rr", ascending=False)
print(df_overall.to_string())

                                                                method_id        rr      time
28  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.5_x_elastic_bm25_0.1  0.564467  0.179312
27  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.4_x_elastic_bm25_0.2  0.561279  0.180053
23  triple_fusion_qdrant_dense_0.3_x_qdrant_sparse_0.6_x_elastic_bm25_0.1  0.560971  0.177657
22  triple_fusion_qdrant_dense_0.3_x_qdrant_sparse_0.5_x_elastic_bm25_0.2  0.560089  0.180253
32  triple_fusion_qdrant_dense_0.5_x_qdrant_sparse_0.4_x_elastic_bm25_0.1  0.558781  0.180205
46                      tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.6  0.557655  0.180557
16  triple_fusion_qdrant_dense_0.2_x_qdrant_sparse_0.6_x_elastic_bm25_0.2  0.555713  0.178954
17  triple_fusion_qdrant_dense_0.2_x_qdrant_sparse_0.7_x_elastic_bm25_0.1  0.555577  0.178116
31  triple_fusion_qdrant_dense_0.5_x_qdrant_sparse_0.3_x_elastic_bm25_0.2  0.554010  0.182489
48                      tuple_fusion_qdrant_dense_0.5_x_qdra

### Overall only tuples

In [7]:
df_overall_tuple_only = df_overall[df_overall["method_id"].str.contains("tuple")]
df_overall_tuple_only

method_id        rr      time
46  tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.6  0.557655  0.180557
48  tuple_fusion_qdrant_dense_0.5_x_qdrant_sparse_0.5  0.553837  0.175547
44  tuple_fusion_qdrant_dense_0.3_x_qdrant_sparse_0.7  0.553140  0.177974
50  tuple_fusion_qdrant_dense_0.6_x_qdrant_sparse_0.4  0.549166  0.178436
42  tuple_fusion_qdrant_dense_0.2_x_qdrant_sparse_0.8  0.547480  0.177617
40  tuple_fusion_qdrant_dense_0.1_x_qdrant_sparse_0.9  0.540570  0.180089
64  tuple_fusion_qdrant_sparse_0.8_x_elastic_bm25_0.2  0.539204  0.108232
65  tuple_fusion_qdrant_sparse_0.9_x_elastic_bm25_0.1  0.538027  0.103789
52  tuple_fusion_qdrant_dense_0.7_x_qdrant_sparse_0.3  0.533588  0.178387
63  tuple_fusion_qdrant_sparse_0.7_x_elastic_bm25_0.3  0.527437  0.107466
54  tuple_fusion_qdrant_dense_0.8_x_qdrant_sparse_0.2  0.515722  0.180365
51   tuple_fusion_qdrant_dense_0.7_x_elastic_bm25_0.3  0.509888  0.180890
53   tuple_fusion_qdrant_dense_0.8_x_elastic_bm25_0.2  0.509144  0.182887
62  tuple_fusion_qdrant_sparse_0.6_x_elastic_bm25_0.4  0.507581  0.107515
56  tuple_fusion_qdrant_dense_0.9_x_qdrant_sparse_0.1  0.496785  0.179473
49   tuple_fusion_qdrant_dense_0.6_x_elastic_bm25_0.4  0.496047  0.181612
55   tuple_fusion_qdrant_dense_0.9_x_elastic_bm25_0.1  0.495631  0.178433
61  tuple_fusion_qdrant_sparse_0.5_x_elastic_bm25_0.5  0.471262  0.109439
47   tuple_fusion_qdrant_dense_0.5_x_elastic_bm25_0.5  0.466742  0.182196
45   tuple_fusion_qdrant_dense_0.4_x_elastic_bm25_0.6  0.409828  0.181401
60  tuple_fusion_qdrant_sparse_0.4_x_elastic_bm25_0.6  0.406983  0.109691
43   tuple_fusion_qdrant_dense_0.3_x_elastic_bm25_0.7  0.367498  0.179968
59  tuple_fusion_qdrant_sparse_0.3_x_elastic_bm25_0.7  0.363467  0.108957
41   tuple_fusion_qdrant_dense_0.2_x_elastic_bm25_0.8  0.331836  0.181757
58  tuple_fusion_qdrant_sparse_0.2_x_elastic_bm25_0.8  0.328936  0.109835
57  tuple_fusion_qdrant_sparse_0.1_x_elastic_bm25_0.9  0.304646  0.114430
39   tuple_fusion_qdrant_dense_0.1_x_elastic_bm25_0.9  0.302853  0.186531

### Overall only triples

In [8]:
df_overall_triple_only = df_overall[df_overall["method_id"].str.contains("triple")]
print(df_overall_triple_only.to_string())

                                                                method_id        rr      time
28  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.5_x_elastic_bm25_0.1  0.564467  0.179312
27  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.4_x_elastic_bm25_0.2  0.561279  0.180053
23  triple_fusion_qdrant_dense_0.3_x_qdrant_sparse_0.6_x_elastic_bm25_0.1  0.560971  0.177657
22  triple_fusion_qdrant_dense_0.3_x_qdrant_sparse_0.5_x_elastic_bm25_0.2  0.560089  0.180253
32  triple_fusion_qdrant_dense_0.5_x_qdrant_sparse_0.4_x_elastic_bm25_0.1  0.558781  0.180205
16  triple_fusion_qdrant_dense_0.2_x_qdrant_sparse_0.6_x_elastic_bm25_0.2  0.555713  0.178954
17  triple_fusion_qdrant_dense_0.2_x_qdrant_sparse_0.7_x_elastic_bm25_0.1  0.555577  0.178116
31  triple_fusion_qdrant_dense_0.5_x_qdrant_sparse_0.3_x_elastic_bm25_0.2  0.554010  0.182489
35  triple_fusion_qdrant_dense_0.6_x_qdrant_sparse_0.3_x_elastic_bm25_0.1  0.549208  0.180948
9   triple_fusion_qdrant_dense_0.1_x_qdrant_sparse_0.7_x_ela

# Overall per category



In [9]:
df_thin_per_category = df[["method_id", "class_category", "rr", "time"]]
df_thin_per_category

method_id  \
0                                    single_elastic_bm25   
1                                    single_elastic_bm25   
2                                    single_elastic_bm25   
3                                    single_elastic_bm25   
4                                    single_elastic_bm25   
...                                                  ...   
99523  tuple_fusion_qdrant_sparse_0.9_x_elastic_bm25_0.1   
99524  tuple_fusion_qdrant_sparse_0.9_x_elastic_bm25_0.1   
99525  tuple_fusion_qdrant_sparse_0.9_x_elastic_bm25_0.1   
99526  tuple_fusion_qdrant_sparse_0.9_x_elastic_bm25_0.1   
99527  tuple_fusion_qdrant_sparse_0.9_x_elastic_bm25_0.1   

                    class_category    rr      time  
0               instances_100_1000  0.00  0.115999  
1               instances_100_1000  0.00  0.069224  
2               instances_100_1000  0.00  0.050538  
3               instances_100_1000  0.00  0.081411  
4               instances_100_1000  0.00  0.255341  
...                            ...   ...       ...  
99523  instances_100000_1000000000  1.00  0.088526  
99524  instances_100000_1000000000  1.00  0.079080  
99525  instances_100000_1000000000  0.50  0.084768  
99526  instances_100000_1000000000  0.25  0.089515  
99527  instances_100000_1000000000  1.00  0.080565  

[99528 rows x 4 columns]

In [10]:
df_overall_per_category = df_thin_per_category.groupby(["method_id", "class_category"], as_index=False).mean().sort_values(by=["class_category","rr"], ascending=False)
print(df_overall_per_category.to_string())

                                                                 method_id               class_category        rr      time
143  triple_fusion_qdrant_dense_0.2_x_qdrant_sparse_0.7_x_elastic_bm25_0.1           instances_100_1000  0.572621  0.181608
191  triple_fusion_qdrant_dense_0.3_x_qdrant_sparse_0.6_x_elastic_bm25_0.1           instances_100_1000  0.572095  0.180832
231  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.5_x_elastic_bm25_0.1           instances_100_1000  0.567319  0.182463
183  triple_fusion_qdrant_dense_0.3_x_qdrant_sparse_0.5_x_elastic_bm25_0.2           instances_100_1000  0.566710  0.184023
135  triple_fusion_qdrant_dense_0.2_x_qdrant_sparse_0.6_x_elastic_bm25_0.2           instances_100_1000  0.565390  0.181989
375                      tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.6           instances_100_1000  0.563104  0.183400
263  triple_fusion_qdrant_dense_0.5_x_qdrant_sparse_0.4_x_elastic_bm25_0.1           instances_100_1000  0.562916  0.183844
87   tri

### Only instance categories

In [11]:
print(df_overall_per_category[df_overall_per_category["class_category"].str.startswith("instances")].to_string())

                                                                 method_id               class_category        rr      time
143  triple_fusion_qdrant_dense_0.2_x_qdrant_sparse_0.7_x_elastic_bm25_0.1           instances_100_1000  0.572621  0.181608
191  triple_fusion_qdrant_dense_0.3_x_qdrant_sparse_0.6_x_elastic_bm25_0.1           instances_100_1000  0.572095  0.180832
231  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.5_x_elastic_bm25_0.1           instances_100_1000  0.567319  0.182463
183  triple_fusion_qdrant_dense_0.3_x_qdrant_sparse_0.5_x_elastic_bm25_0.2           instances_100_1000  0.566710  0.184023
135  triple_fusion_qdrant_dense_0.2_x_qdrant_sparse_0.6_x_elastic_bm25_0.2           instances_100_1000  0.565390  0.181989
375                      tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.6           instances_100_1000  0.563104  0.183400
263  triple_fusion_qdrant_dense_0.5_x_qdrant_sparse_0.4_x_elastic_bm25_0.1           instances_100_1000  0.562916  0.183844
87   tri

### Only ancestors cetegories

In [12]:
print(df_overall_per_category[df_overall_per_category["class_category"].str.startswith("ancestors")].to_string())

                                                                 method_id            class_category        rr      time
9                                                      single_qdrant_dense  ancestors_180_1000000000  0.812500  0.188149
449                      tuple_fusion_qdrant_dense_0.9_x_qdrant_sparse_0.1  ancestors_180_1000000000  0.812500  0.187990
433                      tuple_fusion_qdrant_dense_0.8_x_qdrant_sparse_0.2  ancestors_180_1000000000  0.809295  0.189488
305  triple_fusion_qdrant_dense_0.8_x_qdrant_sparse_0.1_x_elastic_bm25_0.1  ancestors_180_1000000000  0.802885  0.187152
425                       tuple_fusion_qdrant_dense_0.8_x_elastic_bm25_0.2  ancestors_180_1000000000  0.802885  0.194685
441                       tuple_fusion_qdrant_dense_0.9_x_elastic_bm25_0.1  ancestors_180_1000000000  0.802885  0.187971
417                      tuple_fusion_qdrant_dense_0.7_x_qdrant_sparse_0.3  ancestors_180_1000000000  0.794872  0.190102
289  triple_fusion_qdrant_dense_

### Only children

In [13]:
print(df_overall_per_category[df_overall_per_category["class_category"].str.startswith("children")].to_string())

                                                                 method_id              class_category        rr      time
179  triple_fusion_qdrant_dense_0.3_x_qdrant_sparse_0.5_x_elastic_bm25_0.2       children_10000_100000  0.570499  0.181896
219  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.4_x_elastic_bm25_0.2       children_10000_100000  0.566753  0.182131
131  triple_fusion_qdrant_dense_0.2_x_qdrant_sparse_0.6_x_elastic_bm25_0.2       children_10000_100000  0.564696  0.181742
211  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.3_x_elastic_bm25_0.3       children_10000_100000  0.558944  0.180831
227  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.5_x_elastic_bm25_0.1       children_10000_100000  0.558459  0.182583
291  triple_fusion_qdrant_dense_0.7_x_qdrant_sparse_0.1_x_elastic_bm25_0.2       children_10000_100000  0.558155  0.180793
371                      tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.6       children_10000_100000  0.556276  0.181473
299  triple_fusi

# Overall - only mechanic queries

In [14]:
df_only_mech_queries = df[df["query_id"].str.startswith("mech")]
df_only_mech_queries

method_id  user_id  class_id  \
672                                  single_elastic_bm25        0  20825628   
673                                  single_elastic_bm25        0  20825628   
674                                  single_elastic_bm25        0  20825628   
675                                  single_elastic_bm25      101  20825628   
676                                  single_elastic_bm25      101  20825628   
...                                                  ...      ...       ...   
99523  tuple_fusion_qdrant_sparse_0.9_x_elastic_bm25_0.1      114     23442   
99524  tuple_fusion_qdrant_sparse_0.9_x_elastic_bm25_0.1      114     23442   
99525  tuple_fusion_qdrant_sparse_0.9_x_elastic_bm25_0.1      114     23442   
99526  tuple_fusion_qdrant_sparse_0.9_x_elastic_bm25_0.1      114     23442   
99527  tuple_fusion_qdrant_sparse_0.9_x_elastic_bm25_0.1      114     23442   

                      query_id               class_category    rr      time  
672                 mech_label           instances_100_1000  1.00  0.044602  
673                  mech_desc           instances_100_1000  0.00  0.126199  
674            mech_label_desc           instances_100_1000  0.50  0.139791  
675    mech_label_user_query_0           instances_100_1000  0.00  0.112802  
676    mech_label_user_query_1           instances_100_1000  0.00  0.110528  
...                        ...                          ...   ...       ...  
99523  mech_label_user_query_1  instances_100000_1000000000  1.00  0.088526  
99524  mech_label_user_query_2  instances_100000_1000000000  1.00  0.079080  
99525  mech_label_user_query_3  instances_100000_1000000000  0.50  0.084768  
99526  mech_label_user_query_4  instances_100000_1000000000  0.25  0.089515  
99527  mech_label_user_query_5  instances_100000_1000000000  1.00  0.080565  

[55176 rows x 7 columns]

In [15]:
print(df_only_mech_queries[["method_id", "rr", "time"]].groupby(["method_id"], as_index=False).mean().sort_values(by="rr", ascending=False).to_string())


                                                                method_id        rr      time
28  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.5_x_elastic_bm25_0.1  0.832564  0.188197
23  triple_fusion_qdrant_dense_0.3_x_qdrant_sparse_0.6_x_elastic_bm25_0.1  0.831008  0.185811
27  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.4_x_elastic_bm25_0.2  0.829582  0.188750
22  triple_fusion_qdrant_dense_0.3_x_qdrant_sparse_0.5_x_elastic_bm25_0.2  0.827815  0.189693
17  triple_fusion_qdrant_dense_0.2_x_qdrant_sparse_0.7_x_elastic_bm25_0.1  0.826234  0.187460
32  triple_fusion_qdrant_dense_0.5_x_qdrant_sparse_0.4_x_elastic_bm25_0.1  0.825458  0.189945
16  triple_fusion_qdrant_dense_0.2_x_qdrant_sparse_0.6_x_elastic_bm25_0.2  0.823393  0.187249
31  triple_fusion_qdrant_dense_0.5_x_qdrant_sparse_0.3_x_elastic_bm25_0.2  0.820178  0.191231
46                      tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.6  0.819531  0.190523
44                      tuple_fusion_qdrant_dense_0.3_x_qdra

### Only labels

In [16]:
df_only_mech_queries_labels_only = df_only_mech_queries[df_only_mech_queries["query_id"].isin(["mech_label"])]
df_only_mech_queries_labels_only

method_id  user_id  class_id  \
672                                  single_elastic_bm25        0  20825628   
681                                  single_elastic_bm25        0   2066754   
690                                  single_elastic_bm25        0     34038   
699                                  single_elastic_bm25        0   7725634   
708                                  single_elastic_bm25        0    602884   
...                                                  ...      ...       ...   
99435  tuple_fusion_qdrant_sparse_0.9_x_elastic_bm25_0.1        0   8016240   
99444  tuple_fusion_qdrant_sparse_0.9_x_elastic_bm25_0.1        0     23442   
99453  tuple_fusion_qdrant_sparse_0.9_x_elastic_bm25_0.1        0     30022   
99462  tuple_fusion_qdrant_sparse_0.9_x_elastic_bm25_0.1        0    130019   
99471  tuple_fusion_qdrant_sparse_0.9_x_elastic_bm25_0.1        0  28942331   

         query_id               class_category    rr      time  
672    mech_label           instances_100_1000  1.00  0.044602  
681    mech_label         instances_1000_10000  1.00  0.052587  
690    mech_label       instances_10000_100000  0.25  0.042479  
699    mech_label  instances_100000_1000000000  1.00  0.078395  
708    mech_label        children_10000_100000  1.00  0.072098  
...           ...                          ...   ...       ...  
99435  mech_label           instances_100_1000  0.50  0.062902  
99444  mech_label  instances_100000_1000000000  1.00  0.052739  
99453  mech_label         instances_1000_10000  1.00  0.070665  
99462  mech_label       instances_10000_100000  1.00  0.063478  
99471  mech_label           instances_100_1000  1.00  0.094786  

[3696 rows x 7 columns]

In [17]:
print(df_only_mech_queries_labels_only[["method_id", "rr", "time"]].groupby(["method_id"], as_index=False).mean().sort_values(by="rr", ascending=False).to_string())

                                                                method_id        rr      time
23  triple_fusion_qdrant_dense_0.3_x_qdrant_sparse_0.6_x_elastic_bm25_0.1  0.933631  0.131741
17  triple_fusion_qdrant_dense_0.2_x_qdrant_sparse_0.7_x_elastic_bm25_0.1  0.933631  0.135877
40                      tuple_fusion_qdrant_dense_0.1_x_qdrant_sparse_0.9  0.932738  0.137034
2                                                    single_qdrant_sparse  0.932143  0.066319
65                      tuple_fusion_qdrant_sparse_0.9_x_elastic_bm25_0.1  0.923214  0.068255
10  triple_fusion_qdrant_dense_0.1_x_qdrant_sparse_0.8_x_elastic_bm25_0.1  0.923214  0.134718
42                      tuple_fusion_qdrant_dense_0.2_x_qdrant_sparse_0.8  0.920833  0.135943
44                      tuple_fusion_qdrant_dense_0.3_x_qdrant_sparse_0.7  0.918750  0.135366
46                      tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.6  0.917758  0.140172
28  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.5_x_ela

### Only descriptions

In [18]:
df_only_mech_queries_desc_only = df_only_mech_queries[df_only_mech_queries["query_id"].isin(["mech_desc"])]
df_only_mech_queries_desc_only

method_id  user_id  class_id  \
673                                  single_elastic_bm25        0  20825628   
682                                  single_elastic_bm25        0   2066754   
691                                  single_elastic_bm25        0     34038   
700                                  single_elastic_bm25        0   7725634   
709                                  single_elastic_bm25        0    602884   
...                                                  ...      ...       ...   
99436  tuple_fusion_qdrant_sparse_0.9_x_elastic_bm25_0.1        0   8016240   
99445  tuple_fusion_qdrant_sparse_0.9_x_elastic_bm25_0.1        0     23442   
99454  tuple_fusion_qdrant_sparse_0.9_x_elastic_bm25_0.1        0     30022   
99463  tuple_fusion_qdrant_sparse_0.9_x_elastic_bm25_0.1        0    130019   
99472  tuple_fusion_qdrant_sparse_0.9_x_elastic_bm25_0.1        0  28942331   

        query_id               class_category   rr      time  
673    mech_desc           instances_100_1000  0.0  0.126199  
682    mech_desc         instances_1000_10000  0.2  0.086679  
691    mech_desc       instances_10000_100000  0.0  0.078374  
700    mech_desc  instances_100000_1000000000  0.2  0.058000  
709    mech_desc        children_10000_100000  0.0  0.091256  
...          ...                          ...  ...       ...  
99436  mech_desc           instances_100_1000  0.5  0.124134  
99445  mech_desc  instances_100000_1000000000  1.0  0.130381  
99454  mech_desc         instances_1000_10000  1.0  0.082419  
99463  mech_desc       instances_10000_100000  1.0  0.119998  
99472  mech_desc           instances_100_1000  0.2  0.100639  

[3564 rows x 7 columns]

In [19]:
print(df_only_mech_queries_desc_only[["method_id", "rr", "time"]].groupby(["method_id"], as_index=False).mean().sort_values(by="rr", ascending=False).to_string())

                                                                method_id        rr      time
40                      tuple_fusion_qdrant_dense_0.1_x_qdrant_sparse_0.9  0.873457  0.190836
2                                                    single_qdrant_sparse  0.872671  0.089913
10  triple_fusion_qdrant_dense_0.1_x_qdrant_sparse_0.8_x_elastic_bm25_0.1  0.872531  0.189748
65                      tuple_fusion_qdrant_sparse_0.9_x_elastic_bm25_0.1  0.870202  0.113551
42                      tuple_fusion_qdrant_dense_0.2_x_qdrant_sparse_0.8  0.869585  0.188336
17  triple_fusion_qdrant_dense_0.2_x_qdrant_sparse_0.7_x_elastic_bm25_0.1  0.869444  0.188300
44                      tuple_fusion_qdrant_dense_0.3_x_qdrant_sparse_0.7  0.867424  0.189607
46                      tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.6  0.866512  0.191802
23  triple_fusion_qdrant_dense_0.3_x_qdrant_sparse_0.6_x_elastic_bm25_0.1  0.864969  0.185625
28  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.5_x_ela

### Only labels with descriptions

In [20]:
df_only_mech_queries_label_desc_only = df_only_mech_queries[df_only_mech_queries["query_id"].isin(["mech_label_desc"])]
df_only_mech_queries_label_desc_only

method_id  user_id  class_id  \
674                                  single_elastic_bm25        0  20825628   
683                                  single_elastic_bm25        0   2066754   
692                                  single_elastic_bm25        0     34038   
701                                  single_elastic_bm25        0   7725634   
710                                  single_elastic_bm25        0    602884   
...                                                  ...      ...       ...   
99437  tuple_fusion_qdrant_sparse_0.9_x_elastic_bm25_0.1        0   8016240   
99446  tuple_fusion_qdrant_sparse_0.9_x_elastic_bm25_0.1        0     23442   
99455  tuple_fusion_qdrant_sparse_0.9_x_elastic_bm25_0.1        0     30022   
99464  tuple_fusion_qdrant_sparse_0.9_x_elastic_bm25_0.1        0    130019   
99473  tuple_fusion_qdrant_sparse_0.9_x_elastic_bm25_0.1        0  28942331   

              query_id               class_category        rr      time  
674    mech_label_desc           instances_100_1000  0.500000  0.139791  
683    mech_label_desc         instances_1000_10000  0.200000  0.122716  
692    mech_label_desc       instances_10000_100000  0.066667  0.105941  
701    mech_label_desc  instances_100000_1000000000  1.000000  0.093617  
710    mech_label_desc        children_10000_100000  0.090909  0.139669  
...                ...                          ...       ...       ...  
99437  mech_label_desc           instances_100_1000  1.000000  0.137422  
99446  mech_label_desc  instances_100000_1000000000  1.000000  0.145685  
99455  mech_label_desc         instances_1000_10000  1.000000  0.091105  
99464  mech_label_desc       instances_10000_100000  1.000000  0.160666  
99473  mech_label_desc           instances_100_1000  1.000000  0.124050  

[3564 rows x 7 columns]

In [21]:
print(df_only_mech_queries_label_desc_only[["method_id", "rr", "time"]].groupby(["method_id"], as_index=False).mean().sort_values(by="rr", ascending=False).to_string())

                                                                method_id        rr      time
65                      tuple_fusion_qdrant_sparse_0.9_x_elastic_bm25_0.1  1.000000  0.138232
50                      tuple_fusion_qdrant_dense_0.6_x_qdrant_sparse_0.4  1.000000  0.198456
32  triple_fusion_qdrant_dense_0.5_x_qdrant_sparse_0.4_x_elastic_bm25_0.1  1.000000  0.201137
37  triple_fusion_qdrant_dense_0.7_x_qdrant_sparse_0.2_x_elastic_bm25_0.1  1.000000  0.197762
28  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.5_x_elastic_bm25_0.1  1.000000  0.199770
38  triple_fusion_qdrant_dense_0.8_x_qdrant_sparse_0.1_x_elastic_bm25_0.1  1.000000  0.198438
40                      tuple_fusion_qdrant_dense_0.1_x_qdrant_sparse_0.9  1.000000  0.200857
42                      tuple_fusion_qdrant_dense_0.2_x_qdrant_sparse_0.8  1.000000  0.198100
23  triple_fusion_qdrant_dense_0.3_x_qdrant_sparse_0.6_x_elastic_bm25_0.1  1.000000  0.193896
44                      tuple_fusion_qdrant_dense_0.3_x_qdra

### Only labels with user queries

In [22]:
df_only_mech_queries_label_user_only = df_only_mech_queries[~df_only_mech_queries["query_id"].isin(["mech_desc", "mech_label", "mech_label_desc"])]
df_only_mech_queries_label_user_only

method_id  user_id  class_id  \
675                                  single_elastic_bm25      101  20825628   
676                                  single_elastic_bm25      101  20825628   
677                                  single_elastic_bm25      101  20825628   
678                                  single_elastic_bm25      101  20825628   
679                                  single_elastic_bm25      101  20825628   
...                                                  ...      ...       ...   
99523  tuple_fusion_qdrant_sparse_0.9_x_elastic_bm25_0.1      114     23442   
99524  tuple_fusion_qdrant_sparse_0.9_x_elastic_bm25_0.1      114     23442   
99525  tuple_fusion_qdrant_sparse_0.9_x_elastic_bm25_0.1      114     23442   
99526  tuple_fusion_qdrant_sparse_0.9_x_elastic_bm25_0.1      114     23442   
99527  tuple_fusion_qdrant_sparse_0.9_x_elastic_bm25_0.1      114     23442   

                      query_id               class_category    rr      time  
675    mech_label_user_query_0           instances_100_1000  0.00  0.112802  
676    mech_label_user_query_1           instances_100_1000  0.00  0.110528  
677    mech_label_user_query_2           instances_100_1000  0.00  0.072722  
678    mech_label_user_query_3           instances_100_1000  0.00  0.073281  
679    mech_label_user_query_4           instances_100_1000  0.00  0.078270  
...                        ...                          ...   ...       ...  
99523  mech_label_user_query_1  instances_100000_1000000000  1.00  0.088526  
99524  mech_label_user_query_2  instances_100000_1000000000  1.00  0.079080  
99525  mech_label_user_query_3  instances_100000_1000000000  0.50  0.084768  
99526  mech_label_user_query_4  instances_100000_1000000000  0.25  0.089515  
99527  mech_label_user_query_5  instances_100000_1000000000  1.00  0.080565  

[44352 rows x 7 columns]

In [23]:
print(df_only_mech_queries_label_user_only[["method_id", "rr", "time"]].groupby(["method_id"], as_index=False).mean().sort_values(by="rr", ascending=False).to_string())

                                                                method_id        rr      time
27  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.4_x_elastic_bm25_0.2  0.813320  0.191920
28  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.5_x_elastic_bm25_0.1  0.809826  0.191730
22  triple_fusion_qdrant_dense_0.3_x_qdrant_sparse_0.5_x_elastic_bm25_0.2  0.808211  0.193017
31  triple_fusion_qdrant_dense_0.5_x_qdrant_sparse_0.3_x_elastic_bm25_0.2  0.806615  0.194511
23  triple_fusion_qdrant_dense_0.3_x_qdrant_sparse_0.6_x_elastic_bm25_0.1  0.806148  0.189682
32  triple_fusion_qdrant_dense_0.5_x_qdrant_sparse_0.4_x_elastic_bm25_0.1  0.805120  0.193427
16  triple_fusion_qdrant_dense_0.2_x_qdrant_sparse_0.6_x_elastic_bm25_0.2  0.801597  0.190820
17  triple_fusion_qdrant_dense_0.2_x_qdrant_sparse_0.7_x_elastic_bm25_0.1  0.799849  0.190784
26  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.3_x_elastic_bm25_0.3  0.796398  0.194415
48                      tuple_fusion_qdrant_dense_0.5_x_qdra

### Overall per category

In [24]:
df_only_mech_queries_per_category_thin = df_only_mech_queries[["method_id", "class_category", "rr", "time"]]
df_only_mech_queries_per_category_thin

method_id  \
672                                  single_elastic_bm25   
673                                  single_elastic_bm25   
674                                  single_elastic_bm25   
675                                  single_elastic_bm25   
676                                  single_elastic_bm25   
...                                                  ...   
99523  tuple_fusion_qdrant_sparse_0.9_x_elastic_bm25_0.1   
99524  tuple_fusion_qdrant_sparse_0.9_x_elastic_bm25_0.1   
99525  tuple_fusion_qdrant_sparse_0.9_x_elastic_bm25_0.1   
99526  tuple_fusion_qdrant_sparse_0.9_x_elastic_bm25_0.1   
99527  tuple_fusion_qdrant_sparse_0.9_x_elastic_bm25_0.1   

                    class_category    rr      time  
672             instances_100_1000  1.00  0.044602  
673             instances_100_1000  0.00  0.126199  
674             instances_100_1000  0.50  0.139791  
675             instances_100_1000  0.00  0.112802  
676             instances_100_1000  0.00  0.110528  
...                            ...   ...       ...  
99523  instances_100000_1000000000  1.00  0.088526  
99524  instances_100000_1000000000  1.00  0.079080  
99525  instances_100000_1000000000  0.50  0.084768  
99526  instances_100000_1000000000  0.25  0.089515  
99527  instances_100000_1000000000  1.00  0.080565  

[55176 rows x 4 columns]

In [25]:
df_only_mech_queries_overall_per_category = df_only_mech_queries_per_category_thin.groupby(["method_id", "class_category"], as_index=False).mean().sort_values(by=["class_category", "rr"], ascending=False)
print(df_only_mech_queries_overall_per_category.to_string())

                                                                 method_id               class_category        rr      time
143  triple_fusion_qdrant_dense_0.2_x_qdrant_sparse_0.7_x_elastic_bm25_0.1           instances_100_1000  0.821640  0.192958
191  triple_fusion_qdrant_dense_0.3_x_qdrant_sparse_0.6_x_elastic_bm25_0.1           instances_100_1000  0.816461  0.190894
87   triple_fusion_qdrant_dense_0.1_x_qdrant_sparse_0.8_x_elastic_bm25_0.1           instances_100_1000  0.812375  0.194824
79   triple_fusion_qdrant_dense_0.1_x_qdrant_sparse_0.7_x_elastic_bm25_0.2           instances_100_1000  0.811861  0.200376
183  triple_fusion_qdrant_dense_0.3_x_qdrant_sparse_0.5_x_elastic_bm25_0.2           instances_100_1000  0.809848  0.196411
135  triple_fusion_qdrant_dense_0.2_x_qdrant_sparse_0.6_x_elastic_bm25_0.2           instances_100_1000  0.809389  0.192292
231  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.5_x_elastic_bm25_0.1           instances_100_1000  0.809127  0.193531
223  tri

# Overall - only user queries

In [26]:
df_only_user_queries = df[df["query_id"].str.startswith("user")]
df_only_user_queries

method_id  user_id  class_id  \
0                                    single_elastic_bm25        1  20825628   
1                                    single_elastic_bm25        1  20825628   
2                                    single_elastic_bm25        1  20825628   
3                                    single_elastic_bm25        1  20825628   
4                                    single_elastic_bm25        1  20825628   
...                                                  ...      ...       ...   
98687  tuple_fusion_qdrant_sparse_0.9_x_elastic_bm25_0.1       14     23442   
98688  tuple_fusion_qdrant_sparse_0.9_x_elastic_bm25_0.1       14     23442   
98689  tuple_fusion_qdrant_sparse_0.9_x_elastic_bm25_0.1       14     23442   
98690  tuple_fusion_qdrant_sparse_0.9_x_elastic_bm25_0.1       14     23442   
98691  tuple_fusion_qdrant_sparse_0.9_x_elastic_bm25_0.1       14     23442   

           query_id               class_category        rr      time  
0      user_query_0           instances_100_1000  0.000000  0.115999  
1      user_query_1           instances_100_1000  0.000000  0.069224  
2      user_query_2           instances_100_1000  0.000000  0.050538  
3      user_query_3           instances_100_1000  0.000000  0.081411  
4      user_query_4           instances_100_1000  0.000000  0.255341  
...             ...                          ...       ...       ...  
98687  user_query_1  instances_100000_1000000000  0.111111  0.080108  
98688  user_query_2  instances_100000_1000000000  0.200000  0.066064  
98689  user_query_3  instances_100000_1000000000  0.000000  0.083738  
98690  user_query_4  instances_100000_1000000000  0.000000  0.085917  
98691  user_query_5  instances_100000_1000000000  0.000000  0.080715  

[44352 rows x 7 columns]

In [27]:
print(df_only_user_queries[["method_id", "rr", "time"]].groupby(["method_id"], as_index=False).mean().sort_values(by="rr", ascending=False).to_string())

                                                                method_id        rr      time
46                      tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.6  0.231868  0.168158
28  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.5_x_elastic_bm25_0.1  0.230941  0.168258
27  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.4_x_elastic_bm25_0.2  0.227496  0.169233
22  triple_fusion_qdrant_dense_0.3_x_qdrant_sparse_0.5_x_elastic_bm25_0.2  0.227026  0.168511
32  triple_fusion_qdrant_dense_0.5_x_qdrant_sparse_0.4_x_elastic_bm25_0.1  0.227021  0.168088
48                      tuple_fusion_qdrant_dense_0.5_x_qdrant_sparse_0.5  0.226363  0.164035
50                      tuple_fusion_qdrant_dense_0.6_x_qdrant_sparse_0.4  0.225552  0.166644
23  triple_fusion_qdrant_dense_0.3_x_qdrant_sparse_0.6_x_elastic_bm25_0.1  0.225032  0.167513
44                      tuple_fusion_qdrant_dense_0.3_x_qdrant_sparse_0.7  0.224636  0.165357
35  triple_fusion_qdrant_dense_0.6_x_qdrant_sparse_0.3_x_ela

### Overall per category

In [28]:
df_only_user_queries_per_category_thin = df_only_user_queries[["method_id", "class_category", "rr", "time"]]
df_only_user_queries_per_category_thin

method_id  \
0                                    single_elastic_bm25   
1                                    single_elastic_bm25   
2                                    single_elastic_bm25   
3                                    single_elastic_bm25   
4                                    single_elastic_bm25   
...                                                  ...   
98687  tuple_fusion_qdrant_sparse_0.9_x_elastic_bm25_0.1   
98688  tuple_fusion_qdrant_sparse_0.9_x_elastic_bm25_0.1   
98689  tuple_fusion_qdrant_sparse_0.9_x_elastic_bm25_0.1   
98690  tuple_fusion_qdrant_sparse_0.9_x_elastic_bm25_0.1   
98691  tuple_fusion_qdrant_sparse_0.9_x_elastic_bm25_0.1   

                    class_category        rr      time  
0               instances_100_1000  0.000000  0.115999  
1               instances_100_1000  0.000000  0.069224  
2               instances_100_1000  0.000000  0.050538  
3               instances_100_1000  0.000000  0.081411  
4               instances_100_1000  0.000000  0.255341  
...                            ...       ...       ...  
98687  instances_100000_1000000000  0.111111  0.080108  
98688  instances_100000_1000000000  0.200000  0.066064  
98689  instances_100000_1000000000  0.000000  0.083738  
98690  instances_100000_1000000000  0.000000  0.085917  
98691  instances_100000_1000000000  0.000000  0.080715  

[44352 rows x 4 columns]

In [29]:
df_only_user_queries_overall_per_category = df_only_user_queries_per_category_thin.groupby(["method_id", "class_category"], as_index=False).mean().sort_values(by=["class_category", "rr"], ascending=False)
print(df_only_user_queries_overall_per_category.to_string())

                                                                 method_id               class_category        rr      time
191  triple_fusion_qdrant_dense_0.3_x_qdrant_sparse_0.6_x_elastic_bm25_0.1           instances_100_1000  0.266637  0.168255
231  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.5_x_elastic_bm25_0.1           instances_100_1000  0.265058  0.168628
391                      tuple_fusion_qdrant_dense_0.5_x_qdrant_sparse_0.5           instances_100_1000  0.264845  0.163902
375                      tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.6           instances_100_1000  0.263425  0.167922
263  triple_fusion_qdrant_dense_0.5_x_qdrant_sparse_0.4_x_elastic_bm25_0.1           instances_100_1000  0.263178  0.168558
183  triple_fusion_qdrant_dense_0.3_x_qdrant_sparse_0.5_x_elastic_bm25_0.2           instances_100_1000  0.262788  0.168537
287  triple_fusion_qdrant_dense_0.6_x_qdrant_sparse_0.3_x_elastic_bm25_0.1           instances_100_1000  0.262636  0.169576
359     

### Choosing the best match per class of a user

- The idea is that if user writes the 6 queries, and at least one query find the result, the user can use it to model the data structure.
- We want to choose the best query for each of the user assigned classes.
- We can do it by grouping by: method, user, and class.
  - Then we select a maximum for the groups from reciprocal rank.
  - Then the results will be used to again group by a method. 

In [30]:
df_only_user_queries_max_per_user_class = df_only_user_queries[["method_id", "user_id", "class_id", "rr"]].groupby(["method_id", "user_id", "class_id"], as_index=False).max()
df_only_user_queries_max_per_user_class

method_id  user_id  class_id  \
0                                   single_elastic_bm25        1     34038   
1                                   single_elastic_bm25        1    602884   
2                                   single_elastic_bm25        1   2066754   
3                                   single_elastic_bm25        1   4897819   
4                                   single_elastic_bm25        1   7725634   
...                                                 ...      ...       ...   
7387  tuple_fusion_qdrant_sparse_0.9_x_elastic_bm25_0.1       14   8016240   
7388  tuple_fusion_qdrant_sparse_0.9_x_elastic_bm25_0.1       14  11483816   
7389  tuple_fusion_qdrant_sparse_0.9_x_elastic_bm25_0.1       14  18593264   
7390  tuple_fusion_qdrant_sparse_0.9_x_elastic_bm25_0.1       14  28942331   
7391  tuple_fusion_qdrant_sparse_0.9_x_elastic_bm25_0.1       14  51591359   

            rr  
0     0.000000  
1     0.000000  
2     0.000000  
3     0.000000  
4     0.000000  
...        ...  
7387  0.000000  
7388  1.000000  
7389  1.000000  
7390  0.333333  
7391  1.000000  

[7392 rows x 4 columns]

### Average rr with only best ranks per classes of users


In [31]:
print(df_only_user_queries_max_per_user_class[["method_id", "rr"]].groupby(["method_id"], as_index=False).mean().sort_values(by="rr", ascending=False).to_string())

                                                                method_id        rr
46                      tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.6  0.582642
44                      tuple_fusion_qdrant_dense_0.3_x_qdrant_sparse_0.7  0.566155
42                      tuple_fusion_qdrant_dense_0.2_x_qdrant_sparse_0.8  0.560553
28  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.5_x_elastic_bm25_0.1  0.555654
50                      tuple_fusion_qdrant_dense_0.6_x_qdrant_sparse_0.4  0.555480
23  triple_fusion_qdrant_dense_0.3_x_qdrant_sparse_0.6_x_elastic_bm25_0.1  0.553766
48                      tuple_fusion_qdrant_dense_0.5_x_qdrant_sparse_0.5  0.552827
17  triple_fusion_qdrant_dense_0.2_x_qdrant_sparse_0.7_x_elastic_bm25_0.1  0.548075
32  triple_fusion_qdrant_dense_0.5_x_qdrant_sparse_0.4_x_elastic_bm25_0.1  0.546496
27  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.4_x_elastic_bm25_0.2  0.545263
40                      tuple_fusion_qdrant_dense_0.1_x_qdrant_sparse_0.9  0

### Number of classes found per method with only best ranks

- In total we have 56 classes and each class is given to two users.
- Each method have a change of scoring 112 found classes.

In [32]:
df_only_user_queries_max_per_user_class_thin = df_only_user_queries_max_per_user_class[["method_id", "rr"]].copy()
# Round up rr meaning the method found the class.
df_only_user_queries_max_per_user_class_thin["rr"] = df_only_user_queries_max_per_user_class_thin["rr"].apply(lambda x: math.ceil(x))
# Group the results by method -> for a method, each class is presented twice (since two users provided the queries).
print(df_only_user_queries_max_per_user_class_thin.groupby(["method_id"], as_index=False).sum().sort_values(by="rr", ascending=False).to_string())


                                                                method_id  rr
50                      tuple_fusion_qdrant_dense_0.6_x_qdrant_sparse_0.4  98
32  triple_fusion_qdrant_dense_0.5_x_qdrant_sparse_0.4_x_elastic_bm25_0.1  98
28  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.5_x_elastic_bm25_0.1  98
27  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.4_x_elastic_bm25_0.2  98
48                      tuple_fusion_qdrant_dense_0.5_x_qdrant_sparse_0.5  98
35  triple_fusion_qdrant_dense_0.6_x_qdrant_sparse_0.3_x_elastic_bm25_0.1  97
31  triple_fusion_qdrant_dense_0.5_x_qdrant_sparse_0.3_x_elastic_bm25_0.2  97
26  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.3_x_elastic_bm25_0.3  97
16  triple_fusion_qdrant_dense_0.2_x_qdrant_sparse_0.6_x_elastic_bm25_0.2  96
15  triple_fusion_qdrant_dense_0.2_x_qdrant_sparse_0.5_x_elastic_bm25_0.3  96
46                      tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.6  96
44                      tuple_fusion_qdrant_dense_0.3_x_qdrant_s

### For each method and a class we get the number of successful search queries

- There is 12 tries for each method, since we got 2 users and 6 queries per a user.


In [33]:
df1_method_class_successes_thin = df_only_user_queries[["method_id", "class_id", "rr"]]
df1_method_class_successes_thin = df1_method_class_successes_thin.copy()
df1_method_class_successes_thin["rr"] = df1_method_class_successes_thin["rr"].apply(lambda x: math.ceil(x))
print(df1_method_class_successes_thin.groupby(["class_id", "method_id"], as_index=False).sum().to_string())

       class_id                                                              method_id  rr
0           532                                                    single_elastic_bm25   0
1           532                                                    single_qdrant_dense   4
2           532                                                   single_qdrant_sparse   7
3           532  triple_fusion_qdrant_dense_0.1_x_qdrant_sparse_0.1_x_elastic_bm25_0.8   0
4           532  triple_fusion_qdrant_dense_0.1_x_qdrant_sparse_0.2_x_elastic_bm25_0.7   0
5           532  triple_fusion_qdrant_dense_0.1_x_qdrant_sparse_0.3_x_elastic_bm25_0.6   4
6           532  triple_fusion_qdrant_dense_0.1_x_qdrant_sparse_0.4_x_elastic_bm25_0.5   6
7           532  triple_fusion_qdrant_dense_0.1_x_qdrant_sparse_0.5_x_elastic_bm25_0.4   7
8           532  triple_fusion_qdrant_dense_0.1_x_qdrant_sparse_0.6_x_elastic_bm25_0.3   7
9           532  triple_fusion_qdrant_dense_0.1_x_qdrant_sparse_0.7_x_elastic_bm25_0.2   7

- If we sum this per a class
- We have 66 methods (3 base, 27 tuple, 36 triple), the class appears twice and 6 queries per class.

In [34]:
print(df1_method_class_successes_thin[["class_id", "rr"]].groupby(["class_id"], as_index=False).sum().sort_values(by="rr", ascending=False).to_string())

     class_id   rr
36   11483816  693
46   23002039  667
12     130019  665
39   15052846  613
48   29654788  602
28    2819161  591
51   51591359  567
38   13442814  558
25    1478437  555
22    1348305  552
16     294414  490
49   30325611  474
41   18593264  468
29    3007842  460
50   35922124  459
1       11303  456
24    1371150  453
55  124030970  441
27    2319498  430
33    5783996  398
20    1134686  396
42   19389637  390
43   20755546  381
47   28942331  375
45   21191270  370
3       24856  360
17     310899  353
0         532  326
5       34038  314
30    3231690  312
23    1364302  305
31    3621673  305
6       34763  291
19    1076052  290
37   12890393  289
52   63125083  257
4       30022  248
9       47521  228
10      79007  209
53  123523193  197
2       23442  171
8       44539  148
44   20825628  144
21    1195098  141
18     602884  125
35    8016240  113
40   18557321  113
7       43229  110
26    2066754  108
34    7725634   95
14     134556   51
15     17904

### Classes the best methods did not find

- triple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.5_x_elastic_bm25_0.1
- tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.6

In [35]:
# Triple fusion triple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.5_x_elastic_bm25_0.1

df1_triple_best = df_only_user_queries_max_per_user_class[df_only_user_queries_max_per_user_class["method_id"].str.contains("triple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.5_x_elastic_bm25_0.1")][["method_id", "class_id", "rr"]].copy()
df1_triple_best["rr"] = df_only_user_queries_max_per_user_class_thin["rr"].apply(lambda x: math.ceil(x))
df1_triple_best[df1_triple_best["rr"] == 0]

method_id   class_id  rr
3139  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...    4897819   0
3140  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...    7725634   0
3141  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...   18557321   0
3144  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...     133036   0
3145  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...     134556   0
3151  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...  123691918   0
3155  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...    4897819   0
3160  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...     133036   0
3165  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...   20755546   0
3167  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...  123691918   0
3172  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...     179049   0
3203  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...     124714   0
3205  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...    1195098   0
3243  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...    8016240   0

- The class it did not find at all:
    - 4897819 - role
    - 133036 - litography
    - 123691918 - tool

In [36]:
# Tuple fusion tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.6

df1_triple_best = df_only_user_queries_max_per_user_class[df_only_user_queries_max_per_user_class["method_id"].str.contains("tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.6")][["method_id", "class_id", "rr"]].copy()
df1_triple_best["rr"] = df_only_user_queries_max_per_user_class_thin["rr"].apply(lambda x: math.ceil(x))
df1_triple_best[df1_triple_best["rr"] == 0]

method_id   class_id  rr
5153  tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.6     602884   0
5155  tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.6    4897819   0
5156  tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.6    7725634   0
5157  tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.6   18557321   0
5160  tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.6     133036   0
5161  tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.6     134556   0
5167  tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.6  123691918   0
5171  tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.6    4897819   0
5176  tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.6     133036   0
5181  tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.6   20755546   0
5183  tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.6  123691918   0
5188  tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.6     179049   0
5219  tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.6     124714   0
5221  tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.6    1195098   0
5249  tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.6      30022   0
5259  tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.6    8016240   0

- The classes it did not find at all
    - 4897819 - role
    - 133036 - litography
    - 123691918 - tool

Interestingly, they are the same as before. Looking into the queries, we tried them manually, and it did not find them.
Thinking, maybe it was since the things were too abstract, the users have a big problem with generating the queries to.

# The least found classes

In [37]:
df[["class_id", "class_category", "rr"]].groupby(["class_id", "class_category"], as_index=False).mean().sort_values(by="rr", ascending=False)

class_id               class_category        rr
12     130019       instances_10000_100000  0.918775
36   11483816         instances_1000_10000  0.787376
28    2819161           instances_100_1000  0.758244
51   51591359       instances_10000_100000  0.755959
41   18593264  instances_100000_1000000000  0.738710
46   23002039         instances_1000_10000  0.728493
43   20755546     ancestors_180_1000000000  0.706572
49   30325611            ancestors_150_180  0.696442
24    1371150           instances_100_1000  0.693112
50   35922124           instances_100_1000  0.692378
25    1478437         instances_1000_10000  0.681514
27    2319498       instances_10000_100000  0.651949
52   63125083           instances_100_1000  0.651760
45   21191270  instances_100000_1000000000  0.648751
22    1348305       instances_10000_100000  0.639882
38   13442814  instances_100000_1000000000  0.617065
20    1134686       instances_10000_100000  0.607675
42   19389637  instances_100000_1000000000  0.591324
39   15052846           instances_100_1000  0.590057
55  124030970     ancestors_180_1000000000  0.587011
16     294414        children_10000_100000  0.576411
29    3007842         instances_1000_10000  0.570740
48   29654788  instances_100000_1000000000  0.557382
17     310899         instances_1000_10000  0.542551
40   18557321            ancestors_150_180  0.525069
26    2066754         instances_1000_10000  0.522349
47   28942331           instances_100_1000  0.504725
3       24856         instances_1000_10000  0.491895
30    3231690       instances_10000_100000  0.480750
1       11303         instances_1000_10000  0.474634
15     179049       instances_10000_100000  0.473058
34    7725634  instances_100000_1000000000  0.466391
23    1364302           instances_100_1000  0.466142
31    3621673         instances_1000_10000  0.430595
5       34038       instances_10000_100000  0.430293
37   12890393           instances_100_1000  0.422576
6       34763       instances_10000_100000  0.418624
33    5783996       instances_10000_100000  0.412996
18     602884        children_10000_100000  0.409709
4       30022         instances_1000_10000  0.385835
2       23442  instances_100000_1000000000  0.353943
44   20825628           instances_100_1000  0.353012
53  123523193           instances_100_1000  0.338368
0         532  instances_100000_1000000000  0.322076
9       47521  instances_100000_1000000000  0.300023
19    1076052         instances_1000_10000  0.285608
21    1195098       instances_10000_100000  0.278990
10      79007  instances_100000_1000000000  0.267424
54  123691918   children_100000_1000000000  0.258869
8       44539         instances_1000_10000  0.253323
7       43229  instances_100000_1000000000  0.224175
35    8016240           instances_100_1000  0.200401
13     133036           instances_100_1000  0.178040
11     124714       instances_10000_100000  0.176394
14     134556  instances_100000_1000000000  0.160332
32    4897819   children_100000_1000000000  0.100209